proposals/Classes.md

```diff
@@ -0,0 +1,309 @@
+constructor WithDefaultValue() {
+  someInt = 1
+}
+
+constructor WithValue(int value) {
+  someInt = 2
+}
+```
+
+Constructors can take zero (or more) parameters, which are passed into the constructor. Constructors
+as well have an implicit `this` keyword (like all non-static class members), which can be used to
+set the initial state of the constructed class.
+
+
+### `var`: A variable/field
+
+A variable in a class holds some data associated with the *instance* of that class. If a `var` is marked
+as `static`, then the value is associated with the *type* of the class.
+
+#### Defining variables
+
+```
+var<int> someInt
+var<int?> someNullableInt
+var<string> someString
+
+static var<int> someStaticInt
+```
+
+#### Variable initialization
+
+When a class is constructed, all `var`s in the class by default receive the "zero" value of their declared
+type. For nullable types (those ending in `?`), the value is `null`. Otherwise, the `new` constructor (if defined)
+is called. If no `new` constructor is defined on the type of the variable being constructed, the variable **must** be nullable
+**or** filled in by **all** constructors on the class itself.
+
+
+### `function`: A function or method
+
+A function in a class executes some action, either on the instance of the class or globally (if marked `static`). If
+a function is not marked `static`, the `this` implicit value inside the scope of function provides access to the instance of the class.
+
+#### Defining functions
+
+```
+function<void> ReturnsVoid() {
+  this.someInt = 2
+}
+
+function<int> ReturnsInt(int a) {
+  return this.someInt + a + 1
+}
+
+static function<bool> DoSomething() {
+  return true
+}
+```
+
+### `property`: A property
+
+A property defines a named field on the class which executes logic on `get` (and optionally, `set`).
+If a property is not marked `static`, the `this` implicit value inside the scope of property provides access to the instance of the class.
+
+Properties must have a `get`, but `set` is optional (if the property is treated as a read-only or an inferred value).
+
+Inside a `set` block, an implicit `value` value is added with the value specified in the set expression.
+
+Properties **cannot** be of type `void`.
+
+#### Defining properties
+
+```
+property<int> SomeInt {
+  get { return this.someInt }
+}
+
+property<int> SomeOtherInt {
+  get { return this.someInt2 }
+  set { this.someInt2 = value }
+}
+```
+
+
+### `operator`: An operator
+
+An operator defines how an operator (+, -, ==, etc) works over a class of that type.
+
+Operators are always treated as **static** (even though they don't have the `static` keyword).
+
+Binary Operators must always take in two parameters, each a nullable value of the parent class,
+representing the left and right hand side of the expression.
+
+Unary Operators must always take in one parameter, a nullable value of the parent class. 
+
+#### Defining operators
+
+```
+operator plus (SomeClass? first, SomeClass? second) {
```

Comment:
I'm not 100% on the naming and structure here: We could use the actual operator symbol (`+`, `-`), but that might be confusing to read. As well, I made the parameters explicit, but they could be implicit based on the operator type and named `left`, `right` and `value`.